File: internal/sam/sender.go
package sam

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"time"

	"github.com/go-i2p/i2pkeys"
	"github.com/go-i2p/sam3"
)

type SenderConfig struct {
	SAMAddr string
	Dest    string
	Size    int64
	Seed    int64
	ID      string
	Timeout time.Duration
}

func RunSender(cfg SenderConfig, w io.Writer) error {
	res := ResultMsg{
		Type:    MsgResult,
		Role:    "sender",
		Success: false,
	}
	enc := json.NewEncoder(w)

	fail := func(format string, args ...any) error {
		res.Error = fmt.Sprintf(format, args...)
		enc.Encode(res)
		return errors.New(res.Error)
	}

	t0 := time.Now()
	deadline := t0.Add(cfg.Timeout)

	s, err := sam3.NewSAM(cfg.SAMAddr)
	if err != nil {
		return fail("SAM connect: %v", err)
	}
	defer s.Close()

	keys, err := s.NewKeys()
	if err != nil {
		return fail("NewKeys: %v", err)
	}

	session, err := s.NewStreamSession(cfg.ID, keys, streamTunnelOptions())
	if err != nil {
		return fail("NewStreamSession: %v", err)
	}
	defer session.Close()

	destAddr := i2pkeys.I2PAddr(cfg.Dest)

	// Dial с повторными попытками до исчерпания таймаута.
	// DialI2P может вернуть ошибку если LeaseSet ещё не распространился.
	// DialI2P блокируется на conn.Read без deadline — оборачиваем в горутину.
	// Если попытка зависает дольше perAttempt, получаем timeout и идём на retry.
	// Горутина утекает до момента когда i2pd закроет TCP-соединение.
	dialOnce := func(perAttempt time.Duration) (net.Conn, error) {
		type dialRes struct {
			conn net.Conn
			err  error
		}
		ch := make(chan dialRes, 1)
		go func() {
			c, e := session.DialI2P(destAddr)
			ch <- dialRes{c, e}
		}()
		select {
		case r := <-ch:
			return r.conn, r.err
		case <-time.After(perAttempt):
			return nil, fmt.Errorf("DialI2P per-attempt timeout (%v)", perAttempt)
		}
	}

	const perAttempt = 90 * time.Second
	retryInterval := 5 * time.Second
	var conn net.Conn
	for {
		conn, err = dialOnce(perAttempt)
		if err == nil {
			break
		}
		if time.Now().Add(retryInterval).After(deadline) {
			return fail("DialI2P: %v", err)
		}
		time.Sleep(retryInterval)
	}
	defer conn.Close()

	t1 := time.Now()
	res.SetupMs = float64(t1.Sub(t0).Milliseconds())

	r := NewPayloadReader(cfg.Size, cfg.Seed)
	written, _, err := SendPayload(conn, r, cfg.Size)
	t2 := time.Now()

	res.PayloadBytes = written
	if err != nil {
		return fail("SendPayload: %v", err)
	}
	time.Sleep(5 * time.Second)

	res.Success = true
	res.TransferMs = float64(t2.Sub(t1).Milliseconds())
	if res.TransferMs > 0 {
		res.GoodputMbps = (float64(written) * 8 / 1e6) / (res.TransferMs / 1000)
	}
	res.SHA256OK = true

	return enc.Encode(res)
}

File: sam-bench/src/sender.rs
use super::{messages::ResultMsg, payload::PayloadReader, wire::send_payload};
use sam3::{SamClient, SamConn, SamError, StreamSession, SAM_TUNNEL_OPTIONS};
use std::{
    error::Error,
    io::Write,
    thread,
    time::{Duration, Instant},
};

pub struct SenderConfig {
    pub sam_addr: String,
    pub dest: String,
    pub size: i64,
    pub seed: i64,
    pub id: String,
    pub timeout: Duration,
}

pub fn run_sender(cfg: SenderConfig, w: &mut impl Write) -> Result<(), Box<dyn Error>> {
    let mut res = ResultMsg::new("sender");
    let fail =
        |res: &mut ResultMsg, w: &mut dyn Write, msg: String| -> Result<(), Box<dyn Error>> {
            res.error = msg.clone();
            serde_json::to_writer(&mut *w, res)?;
            writeln!(w)?;
            Err(msg.into())
        };

    let t0 = Instant::now();
    let deadline = t0 + cfg.timeout;
    let client = SamClient::connect(&cfg.sam_addr);
    let session = match client.new_transient_stream_session(&cfg.id, SAM_TUNNEL_OPTIONS) {
        Ok(s) => s,
        Err(e) => return fail(&mut res, w, format!("SAM session: {e}")),
    };

    let mut conn = match dial_with_retry(&session, cfg.dest.clone(), deadline) {
        Ok(c) => c,
        Err(e) => return fail(&mut res, w, format!("DialI2P: {e}")),
    };

    let t1 = Instant::now();
    res.setup_ms = t1.duration_since(t0).as_millis() as f64;

    let mut reader = PayloadReader::new(cfg.size, cfg.seed);
    let (written, _) = match send_payload(&mut conn, &mut reader, cfg.size) {
        Ok(v) => v,
        Err(e) => return fail(&mut res, w, format!("SendPayload: {e}")),
    };
    let t2 = Instant::now();
    thread::sleep(Duration::from_secs(5));
    drop(conn);
    drop(session);

    res.payload_bytes = written;
    res.success = true;
    res.sha256_ok = true;
    res.transfer_ms = t2.duration_since(t1).as_millis() as f64;
    if res.transfer_ms > 0.0 {
        res.goodput_mbps = (written as f64 * 8.0 / 1e6) / (res.transfer_ms / 1000.0);
    }
    serde_json::to_writer(&mut *w, &res)?;
    writeln!(w)?;
    Ok(())
}

fn dial_with_retry(
    session: &StreamSession,
    dest: String,
    deadline: Instant,
) -> Result<SamConn, SamError> {
    let per_attempt = Duration::from_secs(90);
    let retry_interval = Duration::from_secs(5);

    loop {
        let attempt_timeout = per_attempt.min(deadline.saturating_duration_since(Instant::now()));
        let last_err = match session.dial_timeout(&dest, attempt_timeout) {
            Ok(conn) => return Ok(conn),
            Err(e) => e,
        };

        if Instant::now() + retry_interval > deadline {
            return Err(last_err);
        }
        thread::sleep(retry_interval);
    }
}
